DOC Improve documentation for LayerNorm

aspenstarss opened the pull request:
In this [commit](https://github.com/pytorch/pytorch/pull/59178/commits/7026995f3ca253fbc19bf511d53f48f861799a4a) and [issue](https://github.com/pytorch/pytorch/pull/59178#issuecomment-897485295), the [Line 134](https://github.com/deniskokarev/pytorch/blob/47e286d024c183cb26a464447b34fde88b80d17d/torch/nn/modules/normalization.py#L134) will overwrite the "embedding" variate which would cause an error when initiating `nn.LayerNorm` function.

I suggest renaming the "embedding" in [Line 133](https://github.com/deniskokarev/pytorch/blob/47e286d024c183cb26a464447b34fde88b80d17d/torch/nn/modules/normalization.py#L133) to "embedding_dim".

The final example is:
```
batch, sentence_length, embedding_dim = 20, 5, 10
embedding = torch.randn(batch, sentence_length, embedding_dim)
layer_norm = nn.LayerNorm(embedding_dim)
```

Fixes #{59178}

aspenstarss committed:
DOC Improve documentation for LayerNorm

In this [commit](https://github.com/pytorch/pytorch/pull/59178/commits/7026995f3ca253fbc19bf511d53f48f861799a4a), the [Line 134](https://github.com/deniskokarev/pytorch/blob/47e286d024c183cb26a464447b34fde88b80d17d/torch/nn/modules/normalization.py#L134) will overwrite the "embedding" variate which would cause an error when initiating `nn.LayerNorm` function.

I suggest renaming the "embedding" in [Line 133](https://github.com/deniskokarev/pytorch/blob/47e286d024c183cb26a464447b34fde88b80d17d/torch/nn/modules/normalization.py#L133) to "embedding_dim".

The final example is:
```
batch, sentence_length, embedding_dim = 20, 5, 10
embedding = torch.randn(batch, sentence_length, embedding_dim)
layer_norm = nn.LayerNorm(embedding_dim)
```

vadimkantorov commented:
I suggest, docs should also specify what is the shape of the used mean/var and affine weights.

jbschlosser reviewed (commented):
Thanks for the contribution! I like @vadimkantorov's suggestion above to add shape information for the learnable affine weights.

Ideally there should be an `Attributes:` section in the docs mentioning that `weight` and `bias` are defined when `elementwise_affine=True` with shape `normalized_shape`. (see [Linear's docs](https://github.com/pytorch/pytorch/blob/eea52b7d471350df0cd3823e10b4add5031f35c4/torch/nn/modules/linear.py#L52-L60) as an example)

aspenstarss committed:
Merge branch 'pytorch:master' into master

aspenstarss committed:
DOC Improve documentation for LayerNorm

add an `Attributes` section in the docs to specify  the shape of `weight` and `bias`

aspenstarss commented:
Greate! I have added an `Attributes` section in the docs to specify the shape of `weight` and `bias` following the [Linear's docs](https://github.com/pytorch/pytorch/blob/eea52b7d471350df0cd3823e10b4add5031f35c4/torch/nn/modules/linear.py#L52-L60).

This is my first-time commit code to an open-source repository. Although I improve documentation only, it's a great process.

Thanks for your reply! @vadimkantorov @jbschlosser

aspenstarss committed:
DOC delete a blank line

delete a blank line that contains whitespace which seems can't pass the automatic testing.

aspenstarss committed:
delete a trailing space

jbschlosser reviewed (approved):
LGTM! Thanks for the fix :)

vadimkantorov commented on torch/nn/modules/normalization.py:
```diff
@@ -123,18 +123,27 @@ class LayerNorm(Module):
         >>> layer_norm(embedding)
+        >>>
         >>> # Image Example
         >>> N, C, H, W = 20, 5, 10, 10
```
Just curious, is this image example realistic? I think almost all the time, LayerNorm is applied over only the embedding dim (and aggregates away the temporal / spatial dims)

aspenstarss replied:
Yes, it is. As the figure mentioned in [here](https://github.com/pytorch/pytorch/issues/51455#issuecomment-771004522), LayerNorm normalizes over the last three (C, H, W) dimensions in compute vision. In my opinion, LayerNorm should normalize over the last two (seq_len, emb_dim) dimensions in NLP following [the original paper](https://arxiv.org/abs/1607.06450). However, the implementation of LayerNorm in the most popular NLP project (such as [AllenNLP](https://github.com/allenai/allennlp/blob/f877fdc30d18178b88c335fbd92722fb77c42d93/allennlp/modules/layer_norm.py#L6) is applied over only the embedding dim.

vadimkantorov replied:
I think the GroupNorm paper isn't a good reference, because LayerNorm seems to be used differently these days (as you mention, usually in NLP only normalising over embedding_dim).

I think for image example, some realistic example is needed from some established / representative / recent model / paper. Otherwise, it is confusing.


Problem with normalising over height and width as well is that the affine parameters are then tied to fixed height and width and won't support other dimensions. I don't think this is the most common scenario. Or at least this should be discussed.
